apps/workshop/projects/Perp88/helpers/calculation.ts

```diff
@@ -0,0 +1,86 @@
+import { Context } from '@defiyield/sandbox';
+import { BigNumber, ethers } from 'ethers';
+import {
+  CHAINLINK_ORACLE_ADDR,
+  COMPOSITION_TOKENS,
+  PLP_STAKING_ADDR,
+  PLP_STAKING_REVENUE_ADDR,
+  PLP_TOKEN_ADDR,
+  POOL_DIAMOND_ADDR,
+  } from './config';
+import GetterFacetAbi from '../abis/GetterFacet.json';
+import FeedableRewarderAbi from '../abis/FeedableRewarder.json';
+import PLPStakingAbi from '../abis/PLPStaking.json';
+import PoolOracleAbi from '../abis/PoolOracle.json';
+import ERC20Abi from '../abis/ERC20.json';
+import { e18, secondsInYear, Zero } from './constant';
+import { FetchPoolsContext } from '../../../../sandbox/src/types/module';
+import { parseEther } from 'ethers/lib/utils';
+
+interface TokenPrice {
+  minPrice: BigNumber;
+  maxPrice: BigNumber;
+  avgPrice: BigNumber;
+  displayPrice: string;
+}
+
+const decimalPlaceDefault = 2;
+export async function calAPR(ctx: FetchPoolsContext): Promise<string> {
+  const { tokens,ethcall, ethcallProvider, logger } = ctx;
+
+  const plpToken = tokens && tokens.find(i=> i.address.toLowerCase()=== PLP_TOKEN_ADDR.toLowerCase())
+  if(!plpToken || !plpToken.price)throw new Error(`PLP Price is invalid`)
+
+  const plpPriceBN = parseEther(plpToken.price.toString())
+
+  const rewardContract = new ethcall.Contract(PLP_STAKING_REVENUE_ADDR, FeedableRewarderAbi);
+  const plpStakingContract = new ethcall.Contract(PLP_STAKING_ADDR, PLPStakingAbi);
+
+  const [rewardRate, totalShare]: BigNumber[] = await ethcallProvider.all([
+    rewardContract.rewardRate(),
+    plpStakingContract.calculateTotalShare(PLP_STAKING_REVENUE_ADDR),
+  ]);
+
+  const rewardRateE18 = rewardRate.mul(10 ** 12); //usdc is 1e6
+
+  const price = await getTokenOraclePrice(ctx, COMPOSITION_TOKENS.USDC);
```

Comment:
We're also providing a price for the USDC token, so you don't need to get it from the oracle

Reply:
is provided price from chainlink?

Reply:
We are getting prices from different places and calculating the average.